### DISCUSSION TITLE
Model development workflow

### DISCUSSION THREAD

**Discussion by jinningwang** [CURENT/andes#241] [General]
# Model Development Workflow in ADNES

ANDES allows users to develop new models by their own will, which can be easily done by translating block diagrams into models.
Before developing a new model, it is recommended to go through the Model References and check out the existing groups and their functions.

1. Make sure you have installed ANDES in development mode.
2. Check existing models that are similar to your model, and then draft the diagram of your model. Some standard model references are available on ESIG, Neplan, PowerWorld, and PSSE Model Library.
3. When implementing the model, similar function pieces from existing models can help to reduce manual efforts.
5. Build your model and finalize it.
7. Re-generate the pycode by ``andes prep``. Sometimes there occur multi-functionalities, where you may need ``andes prep -f`` to re-generate all model codes.
8. Create a test case file that includes your model, where the IEEE 14-bus system and Kundur system can be used.
9. Load the test case, and run the `ss.TDS.init()` to see if the initialization is successful. If initialization runs into an error, you may need to tune the model initialization equations (``v_str`` defined in the model). Remember to run ``andes prep`` and restart the kernel (for Jupyter Notebook users) every time you make a change to the model code.
10. After finishing the model development, it is encouraged to check the code style (can be done by package flake8) and open a PR to contribute it into the ADNES repo.

**Comment by jinningwang** [comment]
Hi there,

Based on my development experience, I summarized this simple guide as a clue for those who want to try ANDES and develop models.

Welcome any suggestions if you think the guide can be enhanced.

Regards,
Jinning

**Reply by cuihantao** [reply]
Thanks a ton!
